### DISCUSSION TITLE
Using `pre-commit` with `postUpgradeTasks`

### DISCUSSION THREAD

**Discussion by adam-moss** [renovatebot/renovate#23856] [Request Help]
### How are you running Renovate?

self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

36

### If you're self-hosting Renovate, select which platform you are using.

gitlab.com

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

> I never saw this working

This is more I've never tried this and an seeking some additional clarification / advice.

I want to run `pre-commit` based hooks on renovate branches.  If the hook fails I want the dependency renovation to fail in a similar way to the "dependency lookup warnings" feature.

### What you tried so far.

Using the official docker image `renovate/renovate:36@sha256:b961d4ad948fa35ca32e778664673645dcb908e5ed77328045ff2e8eebf087ff`

Adding (via env vars for now):

```json
  "allowedPostUpgradeCommands": ["^pip install pre-commit", "^pre-commit install", "^pre-commit run --hook-stage manual"],
  "allowPostUpgradeCommandTemplating": true,
```

and

```json
 "postUpgradeTasks": {
           "commands": [
             "pip install pre-commit",
             "pre-commit install --install-hooks",
             "pre-commit run --hook-stage manual renovate"
           ],
           "executionMode": "branch"
         },
```

Looking in the debug log I can see `pre-commit` is installed successfully but fails to execute as it presumably isn't installed somewhere on `$PATH`

Clarification/advice sought:

1. Where would `pre-commit` be installed to?
2. How can I fail the renovation event if the hook itself fails?
3. When the repository being renovated _isn't_ a python project, how can I ensure python is installed to make `pip` available to install `pre-commit`?
4. If the repository being renovated contains a `.pre-commit-config.yaml` how can I replace this with my own but _not_ commit that changed file?
5. Does a task get access to the full environment or is it in a sanitised subshell like homebrew (i.e. could a task be used to usurp the renovate env vars / tokens)?
6. Is there a better way to achieve this with renovate?

If it makes a difference the config also sets `binarySource: install`

Any guidance greatly appreciated!

Thanks :smile:

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

**Comment by lmilbaum** [comment]
pre-commit can be invoked in the CI

**Reply by adam-moss** [reply]
Yup, but at that point the commit has already been made, so I doesn't meet my needs

**Reply by lmilbaum** [reply]
In that case, you might want to maintain a custom renovate runner image with all the tooling you need for running the pre-commit as a post upgrade task.

**Reply by adam-moss** [reply]
If I do that is that not going to conflict with any python version required by the renovated project?

**Reply by lmilbaum** [reply]
Renovate doesn't trigger python for renovating a python project. I don't see how this kind of conflict can happen.
